### PR TITLE
Integrated Gaussian noise redux

### PR DESCRIPTION
**Pull request by ben18785**
- rebased properly this time, I hope!

### CODE REVIEW

**Comment by ben18785** [comment]
Replaces #745

File: pints/_log_likelihoods.py [hunk 1]
```diff
@@ -133,6 +133,131 @@ def __call__(self, x):
...
+    same estimates as the full `pints.GaussianLogLikelihood`.
+    Arguments:
+    ``problem``
+        A :class:`SingleOutputProblem` or :class:`MultiOutputProblem`.
+    ``lower``
+        The lower limit on the uniform prior om `sigma`. Must be
+        non-negative.
+    ``upper``
+        The upper limit on the uniform prior om `sigma`.
+    *Extends:* :class:`ProblemLogLikelihood`
```
Comment (MichaelClerx) [inline pints/_log_likelihoods.py]: Please add newline before extends

File: pints/_log_likelihoods.py [hunk 2]
```diff
@@ -133,6 +133,131 @@ def __call__(self, x):
...
+            (1 / 2 - n / 2) \text{log}(SSE) +
+            \text{log}\\left[\Gamma((n - 1) / 2, \frac{SSE}{2 b^2}) -
+            \Gamma((n - 1) / 2, \frac{SSE}{2 a^2}) \right]
+    where :math:`\Gamma(u,v)` is the upper incomplete gamma function
+    as defined here: https://en.wikipedia.org/wiki/Incomplete_gamma_function
+    This log-likelihood is inherently a Bayesian method since it
+    assumes a uniform prior on :math:`\sigma\sim U(a,b)`. However
+    using this likelihood in optimisation routines should yield the
+    same estimates as the full `pints.GaussianLogLikelihood`.
+    Arguments:
```
Comment (MichaelClerx) [inline pints/_log_likelihoods.py]: Newline before & after "Arguments:" required to get the rendered formatting right

File: pints/_log_likelihoods.py [hunk 3]
```diff
@@ -133,6 +133,131 @@ def __call__(self, x):
...
+        if np.isscalar(b):
+            b = np.ones(self._no) * float(b)
+        else:
+            b = pints.vector(b)
+            if len(b) != self._no:
+                raise ValueError(
+                    'Upper limit on uniform prior for sigma must be a ' +
+                    ' scalar or a vector of length n_outputs.')
+        if np.any(b <= 0):
+            raise ValueError('Lower limit on uniform prior for sigma ' +
```
Comment (MichaelClerx) [inline pints/_log_likelihoods.py]: Upper?
and maybe strictly positive

File: pints/_log_likelihoods.py [hunk 4]
```diff
@@ -133,6 +133,131 @@ def __call__(self, x):
...
+        diff = b - a
+        if np.any(diff <= 0):
+            raise ValueError('Upper limit on uniform prior for sigma ' +
+                             'must exceed lower limit.')
+        self._a = a
+        self._b = b
+
+        # Pre-calculate
+        n = self._nt
+        self._n_minus_1_over_2 = (n - 1.0) / 2.0
```
Comment (MichaelClerx) [inline pints/_log_likelihoods.py]: Don't need to change this, but the .0 business is no longer required in python 3. We add a "from future import division" at the top of every file to get the same behaviour on python 2. So throughout pints you can just write `1 / 2` and the result will be `0.5`

File: pints/_log_likelihoods.py [hunk 5]
```diff
@@ -133,6 +133,131 @@ def __call__(self, x):
...
+
+    def __call__(self, x):
+        error = self._values - self._problem.evaluate(x)
+        sse = np.sum(error**2, axis=0)
+
+        # Calculate
+        log_temp = np.zeros(len(self._a2))
+        sse = pints.vector(sse)
+        for i, a in enumerate(self._a2):
+            if not a == 0:
```
Comment (MichaelClerx) [inline pints/_log_likelihoods.py]: `if a != 0` ?

File: pints/_log_likelihoods.py [hunk 6]
```diff
@@ -133,6 +133,131 @@ def __call__(self, x):
...
+    Calculates a log-likelihood assuming independent Gaussian-distributed noise
+    at each time point where :math:`\sigma\sim U(a,b)` has been integrated
+    out of the joint posterior of :math:`p(\theta,\sigma|X)`,
+    .. math::
+        \begin{align} p(\theta|X) &= \int_{0}^{\infty} p(\theta, \sigma|X)
+        \mathrm{d}\sigma\\
+        &\propto \int_{0}^{\infty} p(X|\theta, \sigma) p(\theta, \sigma)
+        \mathrm{d}\sigma,\end{align}
+    Note that this is exactly the same statistical model as
+    `pints.GaussianLogLikelihood` with a uniform prior on :math:`\sigma`.
```
[the remark this answers is not in the document]
Comment (ben18785) [reply pints/_log_likelihoods.py]: Cool. Have changed this now

File: pints/_log_likelihoods.py [hunk 7]
```diff
@@ -133,6 +133,131 @@ def __call__(self, x):
...
+    A possible advantage of this log-likelihood compared with using a
+    `pints.GaussianLogLikelihood`, is that it has one fewer
```
[the remark this answers is not in the document]
Comment (ben18785) [reply pints/_log_likelihoods.py]: Changed!

File: pints/_log_likelihoods.py [hunk 8]
```diff
@@ -133,6 +133,131 @@ def __call__(self, x):
...
+    parameter (:math:`sigma`) which may speed up convergence to the
+    posterior distribution, especially for multi-output problems which will
+    have `n_outputs` fewer parameter dimensions.
```
[the remark this answers is not in the document]
Comment (ben18785) [reply pints/_log_likelihoods.py]: Changed!

File: pints/_log_likelihoods.py [hunk 9]
```diff
@@ -133,6 +133,131 @@ def __call__(self, x):
...
+            \text{log}(2 (b - a) \sqrt(2)) +
+            (1 / 2 - n / 2) \text{log}(SSE) +
+            \text{log}\\left[\Gamma((n - 1) / 2, \frac{SSE}{2 b^2}) -
+            \Gamma((n - 1) / 2, \frac{SSE}{2 a^2}) \right]
+    where :math:`\Gamma(u,v)` is the upper incomplete gamma function
+    as defined here: https://en.wikipedia.org/wiki/Incomplete_gamma_function
+    This log-likelihood is inherently a Bayesian method since it
+    assumes a uniform prior on :math:`\sigma\sim U(a,b)`. However
+    using this likelihood in optimisation routines should yield the
+    same estimates as the full `pints.GaussianLogLikelihood`.
```
[the remark this answers is not in the document]
Comment (ben18785) [reply pints/_log_likelihoods.py]: Changed

File: pints/_log_likelihoods.py (hunk 1 above)
Comment (ben18785) [reply pints/_log_likelihoods.py]: Done

File: pints/_log_likelihoods.py (hunk 2 above)
Comment (ben18785) [reply pints/_log_likelihoods.py]: Done

File: pints/_log_likelihoods.py (hunk 3 above)
Comment (ben18785) [reply pints/_log_likelihoods.py]: Good spot! thanks

File: pints/_log_likelihoods.py (hunk 4 above)
Comment (ben18785) [reply pints/_log_likelihoods.py]: Ok, I'm just paranoid, haha!

File: pints/_log_likelihoods.py (hunk 5 above)
Comment (ben18785) [reply pints/_log_likelihoods.py]: Changed!

**Comment by ben18785** [comment]
@MichaelClerx I've also updated the example notebooks readme so that it's alphabetical. I needed to do a bit of sorting of this in order to merge this branch with master.

**Comment by MichaelClerx** [comment]
@ben18785 might be good to check the rendered docs on https://readthedocs.org/projects/pints/ to see if the maths all turned out as you intended?

**Pull request merged**